### PR TITLE
Improve the UI message for PermissionError from log in

### PR DESCRIPTION
**Pull request by yiboliu**
### Reasons for making this change
Suppress extra stack traces for errors occurring when login credentials are incorrect

### Related issues
#2316 

### Screenshots

image
afer commit 4:
image

### Checklist

* [ ] I've added a screenshot of the changes, if this is a frontend change
* [ ] I've added and/or updated tests, if this is a backend change
* [x] I've run the [pre-commit.sh](https://github.com/codalab/codalab-worksheets/blob/master/pre-commit.sh) script
* [ ] I've updated [docs](https://github.com/codalab/codalab-worksheets/tree/master/docs), if needed

### CODE REVIEW

File: codalab/lib/codalab_manager.py [hunk 1]
```diff
@@ -49,6 +49,16 @@
 from codalab.model.worker_model import WorkerModel
 
 
+def exception_handler(exctype, value, traceback, debug_hook=sys.excepthook):
```
Comment (percyliang) [inline codalab/lib/codalab_manager.py]: Why put this in `codalab_manager.py`?

Comment (percyliang) [reply codalab/lib/codalab_manager.py]: Also, this seems like a potentially nice way to handle exceptions, but I wonder if it might be confusing / dangerous, since it has a global effect on the entire codebase?  Also, since we don't handle our other error messages through this mechanism...

Comment (yiboliu) [reply codalab/lib/codalab_manager.py]: This is not a global change. I put it here because we only want to handle the PermissionError in codalab_manager. I tested that PermissionError raised other places will be displayed in the conventional way.

Comment (epicfaace) [reply codalab/lib/codalab_manager.py]: I think it might be nicer to instead just catch PermissionError in cl.py, just like what we do with `KeyboardInterrupt`: https://github.com/codalab/codalab-worksheets/blob/5b83914cf4e73d70a9fe39ce62513c13222c9e1d/codalab/bin/cl.py#L7-L13

Comment (yiboliu) [reply codalab/lib/codalab_manager.py]: > I think it might be nicer to instead just catch PermissionError in cl.py, just like what we do with `KeyboardInterrupt`:
> 
> https://github.com/codalab/codalab-worksheets/blob/5b83914cf4e73d70a9fe39ce62513c13222c9e1d/codalab/bin/cl.py#L7-L13

Do we want to handle all PermissionErrors in this way? I noticed that we have multiple places raising PermissionError.

Comment (epicfaace) [reply codalab/lib/codalab_manager.py]: Hmm, in that case maybe it would be best to create a subclass of PermissionError called `LoginPermissionError` and then only catch `LoginPermissionError`.

Comment (yiboliu) [reply codalab/lib/codalab_manager.py]: The stacktrace doesn't go thru cl.py:
```
Traceback (most recent call last):
  File "/Users/yibol/codalab-worksheets/venv/bin/cl-worker-manager", line 33, in <module>
    sys.exit(load_entry_point('codalab', 'console_scripts', 'cl-worker-manager')())
  File "/Users/yibol/codalab-worksheets/codalab/worker_manager/main.py", line 121, in main
    manager.run_loop()
  File "/Users/yibol/codalab-worksheets/codalab/worker_manager/worker_manager.py", line 125, in run_loop
    self.run_one_iteration()
  File "/Users/yibol/codalab-worksheets/codalab/worker_manager/worker_manager.py", line 161, in run_one_iteration
    'bundles', params={'worksheet': None, 'keywords': keywords, 'include': ['owner']}
  File "/Users/yibol/codalab-worksheets/codalab/client/json_api_client.py", line 18, in wrapper
    return f(*args, **kwargs)
  File "/Users/yibol/codalab-worksheets/codalab/client/json_api_client.py", line 417, in fetch
    query_params=self._pack_params(params),
  File "/Users/yibol/codalab-worksheets/codalab/worker/rest_client.py", line 66, in _make_request
    access_token = self._get_access_token()
  File "/Users/yibol/codalab-worksheets/codalab/lib/codalab_manager.py", line 432, in <lambda>
    lambda: self._authenticate(address, auth_handler),
  File "/Users/yibol/codalab-worksheets/codalab/lib/codalab_manager.py", line 504, in _authenticate
    raise LoginPermissionError("Invalid username or password.")
codalab.common.LoginPermissionError: Invalid username or password.
```
Instead I did this in `codalab-worksheets/codalab/worker_manager/main.py`

Comment (yiboliu) [reply codalab/lib/codalab_manager.py]: Screenshot updated.

File: codalab/server/auth.py [hunk 2]
```diff
@@ -66,4 +66,6 @@ def _make_token_request(self, data):
         except urllib.error.HTTPError as e:
             if e.code == 401:
                 return None
-            raise
+            raise e
+        except urllib.error.URLError:
+            print("Your login info is incorrect. Please double check")
```
Comment (teetone) [inline codalab/server/auth.py]: In your latest screenshot, it looks like both error messages show up:

"Your login info is incorrect. Please double check"
'Login credentials are incorrect. Please double check'

Comment (yiboliu) [reply codalab/server/auth.py]: Yeah, that's because they come from two separate processes. I can definitely remove one of them, but then we won't know where exactly are errors thrown. How do you think?

Comment (teetone) [reply codalab/server/auth.py]: We should only show one of these errors.

File: codalab/worker_manager/worker_manager.py [hunk 3]
```diff
@@ -124,19 +125,24 @@ def run_loop(self):
...
                 traceback.print_exc()
+            except (urllib.error.URLError, JsonApiException, ssl.SSLError) as e:
+                if "[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed" in str(e):
+                    print("Your login info is incorrect. Please try again.")
+                else:
+                    traceback.print_exc()
+
             if self.args.once:
                 break
+            print('Sleeping {} seconds'.format(self.args.sleep_time))
```
Comment (teetone) [inline codalab/worker_manager/worker_manager.py]: We should probably omit this. This message is being shown to the user.

File: codalab/worker_manager/worker_manager.py [hunk 4]
```diff
@@ -124,19 +125,24 @@ def run_loop(self):
...
                 NotFoundError,
-                JsonApiException,
             ):
                 # Sometimes, network errors occur when running the WorkerManager . These are often
                 # transient exceptions, and retrying the command would lead to success---as a result,
                 # we ignore these network-based exceptions (rather than fatally exiting from the
                 # WorkerManager )
                 traceback.print_exc()
+            except (urllib.error.URLError, JsonApiException, ssl.SSLError) as e:
+                if "[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed" in str(e):
```
Comment (teetone) [inline codalab/worker_manager/worker_manager.py]: Is there a cleaner way to check if the user entered incorrect login info? Is it possible to check the type of exception instead?

File: codalab/worker_manager/main.py [hunk 5]
```diff
@@ -118,7 +120,10 @@ def main():
         )
 
     manager = worker_manager_types[args.worker_manager_name](args)
-    manager.run_loop()
+    try:
+        manager.run_loop()
+    except LoginPermissionError:
+        print('All attempts have failed. Please double check your login credentials.')
```
Comment (teetone) [inline codalab/worker_manager/main.py]: I think you mentioned that it will infinitely prompt the user to retry. Can you clarify for what case it will hit this logic?

**Review by teetone — CHANGES_REQUESTED** [review]
Can you add a test for this change maybe in `test_cli`?

File: codalab/worker_manager/worker_manager.py (hunk 4 above)
Comment (yiboliu) [reply codalab/worker_manager/worker_manager.py]: The reason I do this is that in the case if there are other places throwing these types of errors, we don't want to handle in the same way, as the causes could be different.

File: codalab/worker_manager/main.py (hunk 5 above)
Comment (yiboliu) [reply codalab/worker_manager/main.py]: Currently it's not hitting this point. The reason I didn't delete it is that if we are going to limit the number of retries, or give the user an option to quit, then we can use this message. If we are not going to do that, I'll delete it.

Comment (yiboliu) [reply codalab/worker_manager/main.py:126]: If we are not going to do that, I'll delete it.

File: codalab/worker_manager/worker_manager.py (hunk 4 above)
Comment (teetone) [reply codalab/worker_manager/worker_manager.py]: I'm not sure what you mean by same way. Can you just check the exception type instead or is that not possible?

File: tests/cli/test_cli.py [hunk 6]
```diff
@@ -2366,6 +2366,22 @@ def test_nonexistent(ctx):
...
+    password = os.getenv("CODALAB_PASSWORD")
+    del os.environ["CODALAB_USERNAME"]
+    del os.environ["CODALAB_PASSWORD"]
+    _run_command([cl, 'logout'])
+    os.environ["CODALAB_USERNAME"] = username
+    os.environ["CODALAB_PASSWORD"] = "wrongpassword"
+    result = _run_command([cl_worker_manager, '--server=https://worksheets.codalab.org/', 'slurm-batch', '--partition', 'foo'], 1)
+    print("result is " + str(result))
+    sys.stdout.flush()
+    os.environ["CODALAB_USERNAME"] = username
```
Comment (teetone) [inline tests/cli/test_cli.py:2392]: Isn't this already handled inside the cleanup logic?

File: codalab/worker_manager/worker_manager.py (hunk 4 above)
Comment (yiboliu) [reply codalab/worker_manager/worker_manager.py]: It is possible but I mean, if there are somewhere else throw the same type of error due to some other reason, do we still want to show this message? It might be not accurate.

File: tests/cli/test_cli.py (hunk 6 above)
Comment (yiboliu) [reply tests/cli/test_cli.py:2392]: I'm following the example in `test_auth`

File: tests/cli/test_cli.py [hunk 7]
```diff
@@ -2403,10 +2429,17 @@ def test_nonexistent(ctx):
         choices=list(TestModule.modules.keys()) + ['all', 'default'],
         help='Tests to run from: {%(choices)s}',
     )
+    parser.add_argument(
+        '--cl-worker-manager',
```
Comment (epicfaace) [inline tests/cli/test_cli.py:2433]: Why did you add this argument? Seems like it's not used.

File: tests/cli/test_cli.py [hunk 8]
```diff
@@ -2367,6 +2367,32 @@ def test_nonexistent(ctx):
...
+        [
+            cl_worker_manager,
+            '--server=https://worksheets.codalab.org/',
+            'slurm-batch',
+            '--partition',
+            'foo',
+        ],
+        1,
+    )
+    print("result is " + str(result))
```
Comment (epicfaace) [inline tests/cli/test_cli.py:2390]: Should you be trying to check something is equal such as using `check_equals` instead of just printing the result?

File: tests/cli/test_cli.py (hunk 6 above)
Comment (epicfaace) [reply tests/cli/test_cli.py:2392]: Why is the test for this PR specifically targeting `cl-worker-manager`? it might be simpler to just add a new test to `test_auth` with the regular `cl` command.

File: codalab/server/auth.py [hunk 9]
```diff
@@ -6,7 +6,7 @@
 import urllib.request
 import urllib.parse
 import urllib.error
-from codalab.common import URLOPEN_TIMEOUT_SECONDS
+from codalab.common import URLOPEN_TIMEOUT_SECONDS, LoginPermissionError
```
Comment (epicfaace) [inline codalab/server/auth.py:9]: rm import, not used

File: codalab/worker_manager/worker_manager.py [hunk 10]
```diff
@@ -151,17 +152,20 @@ def run_loop(self):
...
+                urllib.error.URLError,
                 JsonApiException,
+                ssl.SSLError,
             ):
                 # Sometimes, network errors occur when running the WorkerManager . These are often
                 # transient exceptions, and retrying the command would lead to success---as a result,
                 # we ignore these network-based exceptions (rather than fatally exiting from the
                 # WorkerManager )
                 traceback.print_exc()
+            except NotFoundError:
```
Comment (epicfaace) [inline codalab/worker_manager/worker_manager.py:166]: Bit confused why you don't check for LoginPermissionError -- what's the difference between that error and NotFoundError in this context?

File: tests/cli/test_cli.py (hunk 6 above)
Comment (epicfaace) [reply tests/cli/test_cli.py:2392]: Also, we need to have a test that checks to ensure that "Your login info is incorrect. Please try again." is printed to the console when the user enters incorrect credentials.
